flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamIterationHead.java

```diff
@@ -119,13 +183,154 @@ protected void cleanup() throws Exception {
+	 * - Upon restoration the logger sorts sliced logs in the same FIFO order and returns an Iterable that
+	 * gives a singular view of the log.
+	 * <p>
+	 * TODO it seems that ListState.clear does not unregister state. We need to put a hook for that.
+	 *
+	 * @param <IN>
+	 */
+	public static class UpstreamLogger<IN> extends AbstractStreamOperator<IN> implements OneInputStreamOperator<IN, IN> {
+
+		private final StreamConfig config;
+
+		private LinkedList<ListState<StreamRecord<IN>>> slicedLog = new LinkedList<>();
+
+		private UpstreamLogger(StreamConfig config) {
+			this.config = config;
+		}
+
+		public void logRecord(StreamRecord<IN> record) throws Exception {
+			if (!slicedLog.isEmpty()) {
+				slicedLog.getLast().add(record);
+			}
+		}
+
+		public void createSlice(String sliceID) throws Exception {
+			ListState<StreamRecord<IN>> nextSlice =
+				getOperatorStateBackend().getOperatorState(new ListStateDescriptor<>(sliceID,
+					config.<StreamRecord<IN>>getTypeSerializerOut(getUserCodeClassloader())));
+			slicedLog.addLast(nextSlice);
+		}
+
+		public void discardSlice() {
+			ListState<StreamRecord<IN>> logToEvict = slicedLog.pollFirst();
+			logToEvict.clear();
+		}
+
+		public Iterable<StreamRecord<IN>> getReplayLog() throws Exception {
+			final List<String> logSlices = new ArrayList<>(getOperatorStateBackend().getRegisteredStateNames());
+			Collections.sort(logSlices, new Comparator<String>() {
+				@Override
+				public int compare(String o1, String o2) {
+					return Long.valueOf(o1).compareTo(Long.valueOf(o2));
+				}
+			});
+
+			final List<Iterator<StreamRecord<IN>>> wrappedIterators = new ArrayList<>();
+			for (String splitID : logSlices) {
+				wrappedIterators.add(getOperatorStateBackend()
+					.getOperatorState(new ListStateDescriptor<>(splitID,
+						config.<StreamRecord<IN>>getTypeSerializerOut(getUserCodeClassloader()))).get().iterator());
+			}
+
+			if (wrappedIterators.size() == 0) {
+				return new Iterable<StreamRecord<IN>>() {
+					@Override
+					public Iterator<StreamRecord<IN>> iterator() {
+						return Collections.emptyListIterator();
+					}
+				};
+			}
+
+			return new Iterable<StreamRecord<IN>>() {
+				@Override
+				public Iterator<StreamRecord<IN>> iterator() {
+
+					return new Iterator<StreamRecord<IN>>() {
+						int indx = 0;
+						Iterator<StreamRecord<IN>> currentIterator = wrappedIterators.get(0);
+
+						@Override
+						public boolean hasNext() {
+							if (!currentIterator.hasNext()) {
+								progressLog();
+							}
+							return currentIterator.hasNext();
+						}
+
+						@Override
+						public StreamRecord<IN> next() {
+							if (!currentIterator.hasNext() && indx < wrappedIterators.size()) {
+								progressLog();
+							}
+							return currentIterator.next();
+						}
+
+						private void progressLog() {
+							while (!currentIterator.hasNext() && ++indx < wrappedIterators.size()) {
```

Comment:
You do a lot of checking against !currentIterator.hasNext() you could probably have it only in progressLog and call that without the if in the other places (maybe rename to progressIfNecessary)